bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/AnnotationActionHandler.java

```diff
@@ -0,0 +1,143 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.automation.module.core.handler;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.smarthome.automation.Action;
+import org.eclipse.smarthome.automation.AnnotatedActions;
+import org.eclipse.smarthome.automation.annotation.ActionInput;
+import org.eclipse.smarthome.automation.handler.ActionHandler;
+import org.eclipse.smarthome.automation.handler.BaseModuleHandler;
+import org.eclipse.smarthome.automation.type.ActionType;
+import org.eclipse.smarthome.automation.type.Input;
+import org.eclipse.smarthome.automation.type.Output;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * ActionHandler which is dynamically created upon annotation on services
+ *
+ * @author Stefan Triller - initial contribution
+ *
+ */
+public class AnnotationActionHandler extends BaseModuleHandler<Action> implements ActionHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(AnnotationActionHandler.class);
+
+    private final Method method;
+    private final ActionType moduleType;
+    private final AnnotatedActions actionProvider;
+
+    public AnnotationActionHandler(Action module, ActionType mt, Method method, AnnotatedActions actionProvider) {
+        super(module);
+
+        this.method = method;
+        this.moduleType = mt;
+        this.actionProvider = actionProvider;
+    }
+
+    @Override
+    public Map<String, Object> execute(Map<String, Object> context) {
+        Map<String, Object> output = new HashMap<>();
+
+        Annotation[][] annots = method.getParameterAnnotations();
+        List<Object> args = new ArrayList<>();
+
+        for (int i = 0; i < annots.length; i++) {
+            Annotation[] annotationsOnParam = annots[i];
+            if (annotationsOnParam != null && annotationsOnParam.length == 1) {
+                if (annotationsOnParam[0] instanceof ActionInput) {
+                    ActionInput inputAnnotation = (ActionInput) annotationsOnParam[0];
+                    // check if the moduleType has a configdescription with this input
+                    if (hasInput(moduleType, inputAnnotation.name())) {
+                        args.add(i, context.get(inputAnnotation.name()));
+                    } else {
+                        logger.error(
+                                "Annotated method defines input '{}' but the module type '{}' does not specify an input with this name.",
+                                inputAnnotation.name(), moduleType);
+                        // TODO: return an empty output here?
+                        return output;
```

Comment:
WRT the comment: If we return `output` at this place, isn't it empty all the time? Nothing has been added to till now.

Reply:
yes, this was just a placeholder to remind myself on whether I really should return an empty output in this error case.